File: crates/e2e-tests/tests/e2e_tests.rs
use nikkadb_client::NikkaType::{TypeString, TypeU8};
use nikkadb_client::{NikkaClient, NikkaTypeWrapper};
use nikkadb_server::server::NikkaServer;
use std::fs::OpenOptions;
use std::io::{Seek, SeekFrom};
use std::thread::sleep;
use std::thread::spawn;
use std::time::Duration;

#[test]
fn element_insertion_test() {
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut db = NikkaClient::with_port(&port);

    db.set("value", "key");
    assert_eq!(db.get::<String>("value"), Some(String::from("key")));
    db.set("key", "value");
    db.set("one", 1);
    assert_eq!(db.get::<u8>("one").unwrap(), 1);
}

#[test]
fn backup_test() {
    setup_files();
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut db = NikkaClient::with_port(&port);

    for _ in 0..200 {
        db.set("key", "value");
    }

    db.create_deque("numbers", TypeU8);
    db.push_first("numbers", NikkaTypeWrapper::NikkaInt(1));
    db.push_last("numbers", NikkaTypeWrapper::NikkaInt(2));

    sleep(Duration::from_secs(1));

    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut db = NikkaClient::with_port(&port);

    assert_eq!(db.get("key"), Some("value".to_string()));
    assert_eq!(db.pop_first("numbers"), Some(1));
    db.set("should be in wal", 12);

    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut db = NikkaClient::with_port(&port);

    assert_eq!(db.get::<u8>("should be in wal"), Some(12));
}

#[test]
fn element_delete_test() {
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut db = NikkaClient::with_port(&port);

    db.set("value", "key");
    db.remove("value");
    assert_eq!(db.get::<String>("value"), None);
}

#[test]
fn transaction_test() {
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut client = NikkaClient::with_port(&port);

    client.begin_transaction();
    client.set("key1", "value").unwrap();
    client.erase_transaction();
    client.set("key2", "value").unwrap();
    client.send_transaction();

    assert_eq!(client.get::<String>("key1"), None);
    assert_eq!(client.get::<String>("key2").unwrap(), "value".to_string());
}

#[test]
fn regex_test() {
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut client = NikkaClient::with_port(&port);

    client.set("alice:bob", "bob");
    client.set("bob:alice", "alice");
    let mut query = client.get_regex("*:*").unwrap();
    let mut real = vec!["alice:bob".to_string(), "bob:alice".to_string()];
    query.sort();
    real.sort();

    assert_eq!(query, real);
}

#[test]
fn clear_test() {
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut client = NikkaClient::with_port(&port);

    client.set("one", "two");
    client.set("three", 3);
    client.clear_database();

    assert_eq!(client.get::<String>("one"), None);
    assert_eq!(client.get::<u8>("three"), None);
}

#[test]
fn deque_test() {
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut client = NikkaClient::with_port(&port);

    client.create_deque("numbers", TypeU8);
    client.push_first("numbers", NikkaTypeWrapper::NikkaInt(1));
    client.push_last("numbers", NikkaTypeWrapper::NikkaInt(2));
    assert_eq!(client.pop_first("numbers").unwrap_or(0), 1);
    assert_eq!(client.pop_last("numbers").unwrap_or(0), 2);
    assert_eq!(client.pop_last("numbers").unwrap_or(0), 0);

    client.create_deque("strings", TypeString);
    client.push_first("strings", NikkaTypeWrapper::NikkaString("one"));
    client.push_last("strings", NikkaTypeWrapper::NikkaString("two"));
    assert_eq!(
        client.pop_first("strings").unwrap_or("0".to_string()),
        "one".to_string()
    );
    assert_eq!(
        client.pop_last("strings").unwrap_or("0".to_string()),
        "two".to_string()
    );
    assert_eq!(
        client.pop_last("strings").unwrap_or("0".to_string()),
        "0".to_string()
    );
}

#[test]
fn transaction_wal_test() {
    setup_files();
    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut client = NikkaClient::with_port(&port);

    client.begin_transaction();
    client.set("key1", "value").unwrap();
    client.erase_transaction();
    client.set("key2", "value").unwrap();
    client.send_transaction();

    assert_eq!(client.get::<String>("key1"), None);
    assert_eq!(client.get::<String>("key2").unwrap(), "value".to_string());

    sleep(Duration::from_secs(1));

    let db = NikkaServer::with_port("0");
    let port = db.get_port().to_string();

    spawn(|| db.run());

    sleep(Duration::from_millis(100));

    let mut client = NikkaClient::with_port(&port);
    assert_eq!(client.get::<String>("key1"), None);
    assert_eq!(client.get::<String>("key2").unwrap(), "value".to_string());
}

fn setup_files() {
    let log = OpenOptions::new().read(true).write(true).open("log.nikka");

    if let Ok(mut log_file) = log {
        log_file.set_len(0);
        log_file.seek(SeekFrom::Start(0));
    }

    let mut backup = OpenOptions::new()
        .read(true)
        .write(true)
        .open("backup.nikka");

    if let Ok(mut backup_file) = backup {
        backup_file.set_len(0);
        backup_file.seek(SeekFrom::Start(0));
    }
}

File: scala/modules/e2e-tests/src/test/scala/nikkadb/tests/E2ETests.scala
package nikkadb.tests

import munit.FunSuite
import nikkadb.client.{NikkaClient, NikkaType, NikkaTypeWrapper}
import nikkadb.server.NikkaServer
import nikkadb.shared.Serializable.given

class E2ETests extends FunSuite:

  private def spawn(): (String) =
    val db = NikkaServer.withPort("0")
    val port = db.serverSocket.getLocalPort.toString
    new Thread(() => db.run()).start()
    Thread.sleep(100)
    port

  test("element insertion") {
    val port = spawn()
    val db = NikkaClient.withPort(port)
    db.setString("value", "key")
    assertEquals(db.getString("value"), Some("key"))
    db.setString("key", "value")
    db.setInt("one", 1)
    assertEquals(db.getInt("one").get, 1.toByte)
  }

  test("backup") {
    val port = spawn()
    val db = NikkaClient.withPort(port)
    for _ <- 0 until 200 do db.setString("key", "value")

    db.createDeque("numbers", NikkaType.TypeInt)
    db.pushFirst("numbers", NikkaTypeWrapper.NikkaInt(1))
    db.pushLast("numbers", NikkaTypeWrapper.NikkaInt(2))

    Thread.sleep(1000)

    val s2 = NikkaServer.withPort("2220")
    new Thread(() => s2.run()).start()
    Thread.sleep(100)

    val db2 = NikkaClient.withPort("2220")
    assertEquals(db2.getString("key"), Some("value"))
    assertEquals(db2.popFirst[Byte]("numbers"), Some(1.toByte))
  }

  test("element delete") {
    val port = spawn()
    val db = NikkaClient.withPort(port)
    db.setString("value", "key")
    db.remove("value")
    assertEquals(db.getString("value"), None)
  }

  test("transaction") {
    val port = spawn()
    val client = NikkaClient.withPort(port)

    client.beginTransaction()
    client.setString("key1", "value")
    client.eraseTransaction()
    client.setString("key2", "value")
    client.sendTransaction()

    assertEquals(client.getString("key1"), None)
    assertEquals(client.getString("key2").get, "value")
  }

  test("regex") {
    val port = spawn()
    val client = NikkaClient.withPort(port)
    client.setString("alice:bob", "bob")
    client.setString("bob:alice", "alice")
    val query = client.getRegex("*:*").sorted
    val real = List("alice:bob", "bob:alice").sorted
    assertEquals(query, real)
  }

  test("clear") {
    val port = spawn()
    val client = NikkaClient.withPort(port)
    client.setString("one", "two")
    client.setInt("three", 3)
    client.clearDatabase()
    assertEquals(client.getString("one"), None)
    assertEquals(client.getInt("three"), None)
  }

  test("deque") {
    val port = spawn()
    val client = NikkaClient.withPort(port)

    client.createDeque("numbers", NikkaType.TypeInt)
    client.pushFirst("numbers", NikkaTypeWrapper.NikkaInt(1))
    client.pushLast("numbers", NikkaTypeWrapper.NikkaInt(2))
    assertEquals(client.popFirst[Byte]("numbers").getOrElse(0.toByte), 1.toByte)
    assertEquals(client.popLast[Byte]("numbers").getOrElse(0.toByte), 2.toByte)
    assertEquals(client.popLast[Byte]("numbers").getOrElse(0.toByte), 0.toByte)

    client.createDeque("strings", NikkaType.TypeString)
    client.pushFirst("strings", NikkaTypeWrapper.NikkaString("one"))
    client.pushLast("strings", NikkaTypeWrapper.NikkaString("two"))
    assertEquals(client.popFirst[String]("strings").getOrElse("0"), "one")
    assertEquals(client.popLast[String]("strings").getOrElse("0"), "two")
    assertEquals(client.popLast[String]("strings").getOrElse("0"), "0")
  }
